src/odin/spare/spare_job.py

```diff
@@ -0,0 +1,383 @@
+    )
+    new_rows = new_df.filter(pl.col("id").is_in(new_and_changed_ids))
+
+    if soft_deletes.is_empty() and new_rows.is_empty():
+        return (prev_df, False)
+    else:
+        return (pl.concat([prev_df, soft_deletes, new_rows], how="vertical"), True)
+
+
+def add_to_left(df: pl.DataFrame, cols: tuple[str, Any, pl.DataType]) -> pl.DataFrame:
+    """
+    Add constant columns to the left of a dataframe
+
+    Each column is a tuple: (name, value, type)
+    """
+    new_cols = [pl.repeat(value, len(df), dtype=dtype).alias(name) for (name, value, dtype) in cols]
+    return pl.concat([pl.select(*new_cols), df], how="horizontal")
+
+
+def run(
+    config: Config,
+    api_endpoint: ApiEndpoint,
+    tmpdir: str,
+) -> None:
+    """
+    Run the job once.
+
+    In a function separate from the job class so it can be tested.
+    """
+    validate_config(config)
+    source = config["source"]
+    target = config["target"]
+    columns = api_endpoint["columns"]
+    parquet_path = os.path.join(target, f"{api_endpoint['name']}.parquet")
+
+    # get new input files
+    paths = spare_s3.list_objects(source)
+    paths = filter(lambda path: api_endpoint["api_path"] in path, paths)
+    # process new files in chronological order (which is also alphabetical order)
+    paths = sorted(paths)
+    if paths == []:
+        # no new files to process
+        return
+
+    df = load_parquet(parquet_path, columns)
+    changes_in_any_file = False
+    for path in paths:
+        with spare_s3.open_file(os.path.join(source, path)) as f:
+            input_df = load_json(f, columns)
+        # add metadata
+        datetime_ms = datetime_from_delta_path_ms(path)
+        input_df = add_to_left(
+            input_df,
+            [
+                ("ROW_TIMESTAMP", datetime_ms, pl.Datetime("ms")),
+                ("ROW_DELETED", False, pl.Boolean()),
+            ],
+        )
+        (df, changes_in_this_file) = merge_new_data(df, input_df, columns, datetime_ms)
+        changes_in_any_file = changes_in_any_file or changes_in_this_file
+
+    if changes_in_any_file:
+        write_parquet(df, parquet_path, tmpdir)
+
+    if "archive" in config:
+        for path in paths:
+            spare_s3.rename_object(
+                os.path.join(source, path), os.path.join(config["archive"], path)
+            )
+
+
+class SpareJob(OdinJob):
+    """Process JSON from Spare (Paratransit) API calls into Parquet files"""
+
+    def __init__(self, config: Config, spare_api_endpoint: ApiEndpoint) -> None:
+        """Create SpareJob instance."""
+        self.config = config
+        self.api_endpoint = spare_api_endpoint
+        self.start_kwargs = {
+            "name": spare_api_endpoint["name"],
+            "api_path": spare_api_endpoint["api_path"],
+        }
+
+    def run(self) -> int:
+        """Entrypoint to run the job."""
+        run(self.config, self.api_endpoint, self.tmpdir)
+        next_run_secs = 1 * 60  # 1min
```

Comment:
I set the frequency pretty high, once per minute, because consumers of this data will care about freshness. Most runs won't do much aside from check whether there's new input (there won't be).

Is this ok?

Reply:
What I meant by "essentially-unreadable" was related to a hypothetical situation. If this job was schedule to run every minute, and would be receiving new data every minute, it would hypothetically be re-writing the output parquet file every minute. This would create a race condition with anything reading the parquet file. Readers of the parquet file would have very frequent read errors, making the file "essentially-unreadable". 

Based on the additional information that has been provided, regarding the dimensions of this dataset, something around a 6 hour (`next_run_secs=21600`) update frequency seems more appropriate. 

The current design goals of Odin are not intended to support sub 1 hour update frequencies.

Reply:
Okay. Eventually I do want to find a way to go more frequently, since that's something OTA has specifically asked for, but every 6 hours should be fine for now and for the medium term. I've changed it to once per 6 hours [d4b79e2](https://github.com/mbta/odin/pull/20/commits/d4b79e28c0eddf42ff75202ab38fd9c05777c506) , and I've made an Asana task to followup later (potentially after the launch) about pushing the frequency higher https://app.asana.com/1/15492006741476/project/1209584755824899/task/1210176270361205 .